p2p/net/conn.go

```diff
@@ -269,28 +269,26 @@ func (c *FormattedConnection) Send(ctx context.Context, m []byte) error {
 	return nil
 }
 
-// SendSock sends a message directly on the socket without waiting for the queue.
-func (c *FormattedConnection) SendSock(m []byte) error {
+// sendSock sends a message directly on the socket without waiting for the queue.
+func (c *FormattedConnection) sendSock(m []byte) error {
 	c.wmtx.Lock()
 	if c.closed {
 		c.wmtx.Unlock()
 		return fmt.Errorf("connection was closed")
 	}
+	c.wmtx.Unlock()
 
-	err := c.deadliner.SetWriteDeadline(time.Now().Add(c.deadline))
-	if err != nil {
+	if err := c.deadliner.SetWriteDeadline(time.Now().Add(c.deadline)); err != nil {
 		return err
 	}
-	_, err = c.w.WriteRecord(m)
-	if err != nil {
-		cerr := c.closeUnlocked()
-		c.wmtx.Unlock()
-		if cerr != ErrAlreadyClosed {
+	if _, err := c.w.WriteRecord(m); err != nil {
+		c.wmtx.Lock()
+		if err := c.closeUnlocked(); err != ErrAlreadyClosed {
```

Comment:
It will probably be cleaner to put the lock inside `c.closeUnlocked`, which isn't safe anyway without it.